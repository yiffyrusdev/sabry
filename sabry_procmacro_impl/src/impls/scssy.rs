use std::path::PathBuf;

use proc_macro2::TokenStream;
use quote::quote;
use sabry_intrnl::scoper::ArbitraryScope;
use syn::{
    parse::{Parse, ParseStream},
    Ident,
};

use super::{ArbitraryStyleBlock, ArbitraryStyleSyntax};

/// Syntax:
/// `#ident(:$syntax)? { $code }`
///
/// `$syntax`: sass/scss
///
/// `$code`: intended to be valid based on $syntax
pub fn scssy_macro_impl(input: TokenStream, source_path: Option<PathBuf>) -> TokenStream {
    let MacroSyntax {
        ident,
        syntax,
        code,
    } = match syn::parse::Parser::parse2(
        |input: ParseStream<'_>| MacroSyntax::parse_syn(input, source_path),
        input,
    ) {
        Ok(m) => m,
        Err(e) => return e.to_compile_error(),
    };

    // quick raffia syntax check
    match ArbitraryScope::from_source(syntax.0, ident.clone(), &code.code) {
        Ok(_) => {}
        Err(e) => return syn::Error::new(code.span, format!("{e:?}")).into_compile_error(),
    }

    let sourcesass = code.to_string();
    let macro_doc = format!("Arbitrary {:?} code declared with `scssy!`. Pretty usable in tandem with `usey!` and `buildy` at build time", syntax);

    quote! {
        #[doc = #macro_doc]
        #[macro_export]
        macro_rules! #ident {
            () => {#sourcesass};
            (syntax) => {#syntax};
        }
    }
}

pub struct MacroSyntax {
    ident: Ident,
    syntax: ArbitraryStyleSyntax,
    code: ArbitraryStyleBlock,
}

impl MacroSyntax {
    pub fn parse_syn(
        input: syn::parse::ParseStream,
        source_path: Option<PathBuf>,
    ) -> syn::Result<Self> {
        let ident = input.parse::<Ident>()?;
        let syntax = input.parse::<ArbitraryStyleSyntax>()?;

        let code = match source_path {
            Some(sp) => ArbitraryStyleBlock::parse_syn(input, Some(sp))?,
            None => input.parse::<ArbitraryStyleBlock>()?,
        };

        Ok(Self {
            ident,
            syntax,
            code,
        })
    }
}

impl Parse for MacroSyntax {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        Self::parse_syn(input, None)
    }
}

#[cfg(test)]
mod test {}
