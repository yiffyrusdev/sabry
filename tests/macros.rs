pub mod sass {
    use sabry::scssy;

    scssy!(mixins_sass:sass "tests/assets/mixin-module.sass");
    scssy!(mixins_inline_sass:sass {"
    @mixin color($color: black)
        @if $color == black
            color: white
            background-color: green
        @else if $color == white
            color: red
            background-color: black
        @else
            color: #c6c6c6
            background-color: #00ffff
        &::after
            content: 'Colored'
            color: $color
            background-color: blue
            &:hover
                transform: scale(2)
        &:hover
            transform: translateX(10%)
"});
    scssy!(module_sass:sass "tests/assets/module-forward.sass");
    scssy!(module_inline_sass:sass {"
@forward 'mixins_a'
@forward 'mixins_c'
"});
    pub(crate) use mixins_inline_sass as mixins_inline_a;
    pub(crate) use mixins_sass as mixins_a;
    pub(crate) use module_inline_sass as module_inline_a;
    pub(crate) use module_sass as module_a;
}

pub mod scss {
    use sabry::scssy;

    scssy!(mixins_scss:scss "tests/assets/mixin-module.scss");
    scssy!(mixins_inline_scss:scss {"
    @mixin surface($color: black){
        @if $color == black{
            color: white;
            background-color: green;
        }@else if $color == white{
            color: red;
            background-color: black;
        }@else{
            color: #c6c6c6;
            background-color: #00ffff;
        }
        &::after{
            content: 'Colored';
            color: $color;
            background-color: blue;
            &:hover{
                transform: scale(2);
            }
        }
        &:hover{
            transform: translateX(10%);
        }
    }
"});
    scssy!(module_scss:scss "tests/assets/module-forward.scss");
    scssy!(module_inline_scss:scss {"
@forward 'mixins_a';
@forward 'mixins_c';
"});

    pub(crate) use mixins_inline_scss as mixins_inline_c;
    pub(crate) use mixins_scss as mixins_c;
    pub(crate) use module_inline_scss as module_inline_c;
    pub(crate) use module_scss as module_c;
}

#[test]
fn ensure_inline_matches_file() {
    assert_eq!(sass::mixins_a!(), sass::mixins_inline_a!());
    assert_eq!(scss::mixins_c!(), scss::mixins_inline_c!());

    assert_eq!(sass::module_a!(), sass::module_inline_a!());
    assert_eq!(scss::module_c!(), scss::module_inline_c!());
}

#[test]
fn syntax_contract() {
    assert_eq!(scss::mixins_inline_c!(syntax), "scss");
    assert_eq!(scss::mixins_c!(syntax), "scss");
    assert_eq!(scss::module_inline_c!(syntax), "scss");
    assert_eq!(scss::module_c!(syntax), "scss");

    assert_eq!(sass::mixins_a!(syntax), "sass");
    assert_eq!(sass::mixins_inline_a!(syntax), "sass");
    assert_eq!(sass::module_a!(syntax), "sass");
    assert_eq!(sass::module_inline_a!(syntax), "sass");
}

#[cfg(test)]
#[cfg(all(feature = "procmacro", feature = "build"))]
mod build_tests {
    use std::fs;

    use insta::{glob, Settings};
    use sabry::{
        buildmagic::SabryBuilder,
        config::{BehavHashCollision, BehavSassModCollision, SabryConfig},
        usey,
    };

    fn gen_config() -> SabryConfig {
        let mut config = SabryConfig::require().unwrap();
        config.css.bundle = Some("tests/sabry_output/bundle.css".into());
        config.css.scopes = Some("tests/sabry_output/scopes".into());
        config.css.minify = false;
        config.css.prelude = Some(vec!["tests/assets/prelude.css".into()]);

        config.sass.prelude = Some(vec![
            "tests/assets/prelude.sass".into(),
            "tests/assets/prelude.scss".into(),
        ]);
        config.sass.intermediate_dir = "tests/sabry_intermediate".into();
        config.sass.scanroot = "tests".into();
        config.sass.module_name_collision = BehavSassModCollision::Merge;

        config.hash.collision = BehavHashCollision::Error;
        config.hash.size = 16;
        config.hash.use_code_size = true;
        config.hash.use_code_text = true;
        config.hash.use_item_names = true;
        config.hash.use_scope_name = true;

        config.lightningcss.targets.android = Some("10".into());
        config.lightningcss.targets.chrome = Some("100".into());
        config.lightningcss.targets.edge = Some("80".into());
        config.lightningcss.targets.firefox = Some("100".into());
        config.lightningcss.targets.ie = Some("8".into());
        config.lightningcss.targets.ios_saf = Some("13.2".into());
        config.lightningcss.targets.opera = Some("80".into());
        config.lightningcss.targets.safari = Some("11".into());
        config.lightningcss.targets.samsung = Some("80".into());

        config
    }

    fn gen_builder() -> SabryBuilder {
        let config = gen_config();
        SabryBuilder::new(config)
    }

    #[test]
    fn compilation_with_buildy() {
        let mut builder = gen_builder();
        builder
            .build(usey!(
                super::sass::mixins_a!(),
                super::sass::module_a!(),
                super::scss::mixins_c!(),
                super::scss::module_c!()
            ))
            .unwrap();

        glob!("sabry_output/**/*.css", |path| {
            let generated = fs::read_to_string(path).unwrap();

            let mut settings = Settings::clone_current();
            settings.set_snapshot_path("sabry_output_snapshots");
            settings.set_prepend_module_to_snapshot(false);
            settings.set_omit_expression(true);
            settings.remove_info();
            settings.bind(|| {
                insta::assert_snapshot!(generated);
            });
        });
    }
}
