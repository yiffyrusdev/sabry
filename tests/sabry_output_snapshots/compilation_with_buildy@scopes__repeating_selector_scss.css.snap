---
source: tests/macros.rs
assertion_line: 172
input_file: tests/sabry_output/scopes/repeating_selector_scss.css
snapshot_kind: text
---
.n9m9Me_zRl6xP8NIGObY_A.el1 {
  color: red;
}

.n9m9Me_zRl6xP8NIGObY_A.el1-mob {
  color: #00f;
}

.n9m9Me_zRl6xP8NIGObY_A.el1__blist {
  list-style: none;
}

.n9m9Me_zRl6xP8NIGObY_A.el1.n9m9Me_zRl6xP8NIGObY_A.sub {
  margin: 0;
}

.n9m9Me_zRl6xP8NIGObY_A.bus .n9m9Me_zRl6xP8NIGObY_A.el1 {
  padding: 0;
}

.n9m9Me_zRl6xP8NIGObY_A.el2 {
  color: #fff;
}

.n9m9Me_zRl6xP8NIGObY_A.el2-mob {
  color: #000;
}

.n9m9Me_zRl6xP8NIGObY_A.el2__blist {
  background: #fff;
}

.n9m9Me_zRl6xP8NIGObY_A.el2__blist-mob {
  background: #000;
}

.n9m9Me_zRl6xP8NIGObY_A.el2.n9m9Me_zRl6xP8NIGObY_A.sub {
  margin: 0;
}

.n9m9Me_zRl6xP8NIGObY_A.el2 .n9m9Me_zRl6xP8NIGObY_A.bus {
  margin: 0;
}
